program/src/processor.rs

```diff
@@ -210,9 +232,124 @@ fn process_distribute_rewards(
 /// [InitializeHolderRewards](enum.PaladinRewardsInstruction.html)
 /// instruction.
 fn process_initialize_holder_rewards(
-    _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
 ) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let holder_rewards_pool_info = next_account_info(accounts_iter)?;
+    let holder_rewards_info = next_account_info(accounts_iter)?;
+    let token_account_info = next_account_info(accounts_iter)?;
+    let mint_info = next_account_info(accounts_iter)?;
+    let _system_program = next_account_info(accounts_iter)?;
+
+    let token_supply = {
+        let mint_data = mint_info.try_borrow_data()?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+        mint.base.supply
+    };
+
+    let token_account_balance = {
+        let token_account_data = token_account_info.try_borrow_data()?;
+        let token_account = StateWithExtensions::<Account>::unpack(&token_account_data)?;
+
+        // Ensure the provided token account is for the mint.
+        if !token_account.base.mint.eq(mint_info.key) {
+            return Err(PaladinRewardsError::TokenAccountMintMismatch.into());
+        }
+
+        token_account.base.amount
+    };
+
+    let last_seen_total_rewards = {
+        // Ensure the holder rewards pool is owned by the Paladin Rewards
+        // program.
+        if !holder_rewards_pool_info.owner.eq(program_id) {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        // Ensure the provided holder rewards pool address is the correct
+        // address derived from the mint.
+        if !holder_rewards_pool_info
+            .key
+            .eq(&get_holder_rewards_pool_address(mint_info.key))
```

Comment:
This can be done in follow-up work, but I'm just realizing that `get_holder_rewards_pool_address` and  `get_holder_rewards_address` don't take in the `program_id`, which would be more flexible

Reply:
My main test is whether other people might deploy their own. There's only one BPF loader / address lookup table program, and we want to actively *discourage* people from deploying their own versions of these programs. But in this case, others might want to fork the program.

Reply:
Made a note to PR this later.

Reply:
Wouldn't they just update `declare_id!`, though?

Reply:
Maybe, but it's not 100% that they will. Even so, including the program id also allows a program to become an interface so that anyone can call into it. All in all, it's a small change that makes the code agnostic to the declared id, which is what we should generally aim to do.